redis/cluster.py

```diff
@@ -0,0 +1,1249 @@
+            pass
+
+    def process_result(self, command, res, **kwargs):
+        """
+        `res` is a dict with the following structure:
+            Dict(NodeName, CommandResult)
+        """
+        if not isinstance(res, dict):
+            raise ValueError('The result should be of dict type')
+
+        if command in self.result_callbacks:
+            return self.result_callbacks[command](command, res, **kwargs)
+        elif len(res) == 1:
+            # When we execute the command on a single node, we can
+            # remove the dictionary and return a single response
+            return list(res.values())[0]
+        else:
+            return res
+
+
+class ClusterNode(object):
+    def __init__(self, host, port, server_type=None, redis_connection=None):
+        if host == "localhost":
+            host = socket.gethostbyname(host)
+
+        self.host = host
+        self.port = port
+        self.name = get_node_name(host, port)
+        self.server_type = server_type
+        self.redis_connection = redis_connection
+
+    def __repr__(self):
+        return 'host={0},port={1},' \
+               'name={2},server_type={3}' \
+            .format(self.host,
+                    self.port,
+                    self.name,
+                    self.server_type)
+
+    def __eq__(self, obj):
+        return isinstance(obj, ClusterNode) and obj.name == self.name
+
+
+class LoadBalancer:
+    """
+    Round-Robin Load Balancing
+    """
+
+    def __init__(self, bucket_size, start_index=0):
+        self.slot_idx_list = [start_index] * bucket_size
+
+    def get_server_index(self, slot, list_size):
+        server_index = self.slot_idx_list[slot]
+        # Update the index
+        self.slot_idx_list[slot] = (server_index + 1) % list_size
+        return server_index
+
+
+class NodesManager:
+    def __init__(self, startup_nodes=None, from_url=False,
+                 skip_full_coverage_check=False, **kwargs):
+        self.nodes_cache = {}
+        self.slots_cache = {}
+        self.startup_nodes = {}
+        self.populate_startup_nodes(startup_nodes)
+        self.from_url = from_url
+        self._skip_full_coverage_check = skip_full_coverage_check
+        self.reset_connections = False
+        self._moved_exception = None
+        self.connection_kwargs = kwargs
+        self.read_load_balancer = None
+        # Currently we do not support passing connection classes,
+        # we use ClusterConnection by default or SSLClusterConnection for SSL
+        self.initialize()
+
+    def get_node(self, node_name=None, host=None, port=None):
+        if node_name is None:
+            if not host or not port:
+                warnings.warn(
+                    "get_node requires one of the followings: "
+                    "1. node name "
+                    "2. host and port"
+                )
+                return None
+            if host == "localhost":
```

Comment:
elif host is not None and node_name is None

Reply:
changed